echopype/echodata/simrad.py

```diff
@@ -0,0 +1,233 @@
+"""
+Contains functions that are specific to Simrad echo sounders
+"""
+from typing import Optional, Tuple
+
+from .echodata import EchoData
+
+
+def _check_mode_input_without_data(
+    waveform_mode: str, encode_mode: str, pulse_compression: bool
+) -> None:
+    """
+    Checks that the ``waveform_mode`` and ``encode_mode`` have
+    the correct values and that the pair of inputs are valid, without
+    considering the actual data.
+
+    Parameters
+    ----------
+    waveform_mode: str
+        Type of transmit waveform
+    encode_mode: str
+        Type of encoded return echo data
+    pulse_compression: bool
+        States whether pulse compression should be used
+    """
+
+    if waveform_mode not in ["CW", "BB"]:
+        raise RuntimeError("The input waveform_mode must be either 'CW' or 'BB'!")
+
+    if encode_mode not in ["complex", "power"]:
+        raise RuntimeError("The input encode_mode must be either 'complex' or 'power'!")
+
+    # BB has complex data only, but CW can have complex or power data
+    if (waveform_mode == "BB") and (encode_mode == "power"):
+        raise RuntimeError("encode_mode='power' not allowed when waveform_mode='BB'!")
+
+    # make sure that we have BB and complex inputs, if pulse compression is selected
+    if pulse_compression and ((waveform_mode != "BB") or (encode_mode != "complex")):
+        raise RuntimeError(
+            "Pulse compression can only be used with "
+            "waveform_mode='BB' and encode_mode='complex'"
+        )
+
+
+def _check_mode_input_with_data_EK60(
+    echodata: EchoData, waveform_mode: str, encode_mode: str
+) -> Optional[str]:
+    """
+    Ensures that the provided ``waveform_mode`` and ``encode_mode`` are consistent
+    with the EK60-like data supplied by ``echodata``.
+
+    Parameters
+    ----------
+    echodata: EchoData
+        An ``EchoData`` object holding the data
+    waveform_mode : {"CW", "BB"}
+        Type of transmit waveform
+    encode_mode : {"complex", "power"}
+        Type of encoded return echo data
+
+    Returns
+    -------
+    power_ed_group: str, optional
+        The ``EchoData`` beam group path containing the power data
+    """
+
+    # initialize power EchoData group value
+    power_ed_group = None
+
+    # EK60-like sensors must have 'power' and 'CW' modes only
+    if waveform_mode != "CW":
+        raise RuntimeError("Incorrect waveform_mode input provided!")
+    if encode_mode != "power":
+        raise RuntimeError("Incorrect encode_mode input provided!")
+
+    # ensure that no complex data exists (this should never be triggered)
+    # TODO: is this necessary? I just wanted to provide a check against the data ...
+    if "backscatter_i" in echodata["Sonar/Beam_group1"].variables:
```

Comment:
I agree it is a good idea... in theory no one would try to "mock" an EchoData object except for us, but I can see us or others maybe mess up when trying to manipilate an existing EchoData object ??

Reply:
Yes, I was putting it there primarily for a sanity check. I will keep this check in and remove the TODO. Thank you for your input.